CMakeLists.txt

```diff
@@ -88,18 +88,39 @@ cmake_dependent_option( UMPIRE_ENABLE_OPENMP "Build Umpire with OpenMP" On
                        "ENABLE_OPENMP" Off )
 cmake_dependent_option( UMPIRE_ENABLE_MPI "Build Umpire with MPI" On
                        "ENABLE_MPI" Off )
-#
-# NOTE: The following option depends upon UMPIRE_ENABLE_MPI being defined first
-#
-cmake_dependent_option( UMPIRE_ENABLE_IPC_SHARED_MEMORY "Enable Host Shared Memory Resource" ${UMPIRE_ENABLE_MPI}
-                       "NOT WIN32;NOT APPLE" Off )
 
-cmake_dependent_option( UMPIRE_ENABLE_MPI3_SHARED_MEMORY "Enable MPI3 Shared Memory Resource" Off "${ENABLE_MPI}" Off)
+########
+# Shared Memory Options
+########
 
-if (UMPIRE_ENABLE_IPC_SHARED_MEMORY AND UMPIRE_ENABLE_MPI3_SHARED_MEMORY)
-  message(FATAL_ERROR "Please select one shared memory option"
-        "(UMPIRE_ENABLE_IPC_SHARED_MEMORY OR UMPIRE_ENABLE_MPI3_SHARED_MEMORY)")
-endif ()
+option( UMPIRE_ENABLE_IPC_SHARED_MEMORY "Enable Host Shared Memory Resource" Off) 
+
+if(UMPIRE_ENABLE_IPC_SHARED_MEMORY)
+  if(APPLE OR WIN32)
+    message(FATAL_ERROR "IPC_SHARED_MEMORY cannot be enabled on APPLE or WIN32.")
+  endif()
+endif()
+
+option( UMPIRE_ENABLE_MPI3_SHARED_MEMORY "Enable MPI3 Shared Memory Resource" Off)
+
+if(UMPIRE_ENABLE_MPI3_SHARED_MEMORY)
+  if(NOT ENABLE_MPI)
+    message(FATAL_ERROR "UMPIRE_ENABLE_MPI3_SHARED_MEMORY requires ENABLE_MPI to be ON")
+  endif()
+endif()
+
+if(NOT DEFINED UMPIRE_DEFAULT_SHARED_MEMORY_RESOURCE)
+  if(UMPIRE_ENABLE_IPC_SHARED_MEMORY AND NOT UMPIRE_ENABLE_MPI3_SHARED_MEMORY)
+    set(UMPIRE_DEFAULT_SHARED_MEMORY_RESOURCE "POSIX" CACHE STRING "Default implementation for SHARED memory resource (POSIX)")
+  elseif(UMPIRE_ENABLE_MPI3_SHARED_MEMORY AND NOT UMPIRE_ENABLE_IPC_SHARED_MEMORY)
+    set(UMPIRE_DEFAULT_SHARED_MEMORY_RESOURCE "MPI3" CACHE STRING "Default implementation for SHARED memory resource (MPI3)")
+  else()
+    # Either both or neither enabled - default to POSIX
+    set(UMPIRE_DEFAULT_SHARED_MEMORY_RESOURCE "MPI3" CACHE STRING "Default implementation for SHARED memory resource (MPI3)")
```

Comment:
When this wasn't the default, the allocator in the mpi3 shared memory tests would call the "allocate" function that belonged to IPC shared memory, causing a failure because no named was passed in...

Reply:
In this case, it shouldn't be an option if it can't be changed. But I suspect there is a problem with the logic if it can't be changed.